### ISSUE TITLE
Why my validation loss is not converging over multistream model?

### ISSUE THREAD

**Issue by anamika06jain** [keras-team/keras#12031]
I want to merge two CNNs that are trained over the different dataset. I have taken two sequential models and merged them. But when using customized fit_generator, validation loss is not converging. How will I pass generators of different datasets?

    input1_1 = keras.layers.Input(shape=(129,129,3))
    x1 = keras.layers.Conv2D(kernel_size = (3,3), filters = 32, 
    activation='PReLU')(input1_1)

    x3 = keras.layers.MaxPooling2D(2,2)(x1)
    x4 = keras.layers.Conv2D(kernel_size = (5,5), filters = 64, 
     activation='relu')(x3)
    x5 = keras.layers.MaxPooling2D(2,2)(x4)
    x6 = keras.layers.Conv2D(kernel_size = (7,7), filters = 128, 
    activation='relu')(x5)
    d1_1 = keras.layers.Dropout(0.5)(x6)
    br1_1= keras.layers.MaxPooling2D(2,2)(d1_1)
    br1_1 = keras.layers.Flatten()(br1_1)

    input2_2 = keras.layers.Input(shape=(129,129,3))
    x1 = keras.layers.Conv2D(kernel_size = (3,3), filters = 32, 
    activation='PReLU')(input2_2)

    x3 = keras.layers.MaxPooling2D(2,2)(x1)
    x4 = keras.layers.Conv2D(kernel_size = (5,5), filters = 64, 
    activation='relu')(x3)
    x5 = keras.layers.MaxPooling2D(2,2)(x4)
    x6 = keras.layers.Conv2D(kernel_size = (7,7), filters = 128, 
     activation='relu')(x5)
    d2_2 = keras.layers.Dropout(0.5)(x6)
    br2_2= keras.layers.MaxPooling2D(2,2)(d2_2)
    br2_2 = keras.layers.Flatten()(br2_2)

    added1_1 = keras.layers.concatenate([br1_1, br2_2], axis=1)
    d2_3 = keras.layers.Dropout(0.5)(added1_1)
    # d2_4 = keras.layers.Dropout(0.4)(d2_3)
    out1_1 = keras.layers.Dense(159,activation='softmax',kernel_regularizer=regularizers.l2(0.01),
            activity_regularizer=regularizers.l1(0.01))(d2_3)
    # model=keras.layers.Conv2DTranspose(kernel_size= (4,4), filters=10, activation='relu')(out)
    modal1_1 = keras.models.Model(inputs=[input1_1,input2_2], outputs=out1_1)
    modal1_1.summary()

I have tried to zip the generators(train_face, train_iris), but it is not doing the task.

    def custom_iterator(Xp, Xs):

      from keras.preprocessing.image import ImageDataGenerator

      ig1 = ImageDataGenerator(rescale=1./255)
      ig2 = ImageDataGenerator(rescale=1./255)
      temp1 = ig1.flow_from_directory(Xp,target_size = (129, 129),batch_size = 
          10,class_mode = "categorical")
      temp2 = ig2.flow_from_directory(Xs,target_size = (129, 129),batch_size = 
          10,class_mode = "categorical")

      for batch in zip(temp1,temp2):

        yield [batch[0][0], batch[1][0]], [batch[0][1]]

After performing Zip if i call fit generator my validation loss is not converging.

    train_data_dir = "C:\\Users\\Desktop\\SDUMLA\\faceDataset\\mixTrain"
    validation_data_dir = "C:\\Users\\Desktop\\SDUMLA\\faceDataset\\mixTest"
    train_data_dir_nir = "C:\\Users\\Desktop\\SDUMLA\\IRIS(129_129)\\mixTrain"
    validation_data_dir_nir = "C:\\Users\\Desktop\\SDUMLA\\IRIS(129_129)\\mixTest"
    train_gen=custom_iterator(train_data_dir_nir, train_data_dir)
    valid_gen=custom_iterator(validation_data_dir_nir, validation_data_dir)
    nb_train_samples=2226
    nb_validation_samples=1272
    batch_size=10
    Spe=nb_train_samples/ batch_size
    valiStep=nb_validation_samples / batch_size

    print ('compiling')
    modal1_1.compile(optimizer=keras.optimizers.Adam(lr=0.0001, beta_1=0.9, beta_2=0.999, 
          epsilon=None, decay=0.0, amsgrad=False),
          loss='categorical_crossentropy',
          metrics=['accuracy'])

    hist=modal1_1.fit_generator(train_gen,
           steps_per_epoch = Spe,
           epochs = 200,verbose = 1,
           validation_data = valid_gen,   validation_steps = valiStep)

![image](https://user-images.githubusercontent.com/37176621/51083123-b4c71a00-173a-11e9-8fe6-e8f8ce9b8335.png)

   Have I written custom code (as opposed to using example directory): for generators, i have written custom code, that was available on some blog.
OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
TensorFlow backend (yes / no): yes
TensorFlow version (CPU / GPU): GPU (1.8.0)

Python version: 3.5.5
CUDA/cuDNN version: Cuda compilation tools, release 9.1, V9.1.85
GPU model and memory (Full Specifications): NVIDIA GeForce GT 730 2GB, compute capability: 3.5     

how can i deal with this problem? i have added dropouts to reduce overfitting. i don't know what mistake i am doing.

**Comment by msymp** [comment]
Hello @anamika06jain  , can you please provide us your system details in the below template.

System information:

- Have I written custom code (as opposed to using example directory):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow backend (yes / no):
- TensorFlow version (CPU / GPU):
- Python version:
- CUDA/cuDNN version:
- GPU model and memory (Full Specifications):

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

Describe the problem:
Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or a feature request.

Source code / logs:
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by msymp** [comment]
@ParikhKadam , this query from @anamika06jain concerns the merging of models with _fit_generator_ , but doesn't lead to convergence. Can you advise? Thanks.

**Comment by ParikhKadam** [comment]
@anamika06jain @msymp 

I see that the model is not improving itself or may it can't given this type of architecture. The problem should lie in the logic you applied to solve the problem. In order to know what is actually happening, I would need the following details:

1. The problem you are trying to solve
2. The website/book you referred to implement this model.
3. Summary of the model
4. Dataset (along with it's format) you are using to solve the problem

**Comment by ParikhKadam** [comment]
@anamika06jain 

It's definitely very hard to trace.. I can help if you can upload your data and code in kaggle, create a notebook and share it with me.

For now, I can say:
1. Try playing with model's hyperparameters. I think the learning rate is too low, can't tell anything about other parameters of the loss function.
2. Trying changing relu to leakyrelu.
3. There is some problem in your data generator. Look at the training of your model. I can find something like 223/222. How can that be possible?

Something is wrong.. Start printing tensor values, shapes, number of batches, batch index, etc. at runtime in order to fix the generator.

**Comment by anamika06jain** [comment]
@ParikhKadam how can i reach u on kaggle? please tell me your name on kaggle.

**Comment by ParikhKadam** [comment]
@anamika06jain https://www.kaggle.com/kadamparikh

**Comment by anamika06jain** [comment]
@ParikhKadam  i have shared a notebook on kaggle. can you please share your contact mail id ?

**Comment by ParikhKadam** [comment]
@anamika06jain I tired accessing the notebook once but kaggle wasn't assigning me a new kernel for that. So, I just left that out.

Now, when I got access to the kernel, I tried to run the code once but it seems like you haven't uploaded the data or maybe I don't have access to it.

**Comment by ParikhKadam** [comment]
@anamika06jain I don't know why is Keras getting 223th batch of 222 batches generated. There seems a fault in generator and that's the reason I needed the kaggle notebook. But I can't crosscheck my piece of code if I don't have the data.

And it also seems like your model is overfitting on training data.

**Comment by anamika06jain** [comment]
@ParikhKadam i have resolved the problem by switching to fit rather than fit_generator.

**Comment by jvishnuvardhan** [comment]
As it was resolved, I am closing the issue. Please open new ticket if you see similar issue again. Thanks!
